### DISCUSSION TITLE
Get root entry from matrix in matrix

### DISCUSSION THREAD

**Discussion by shifuma** [craftcms/cms#15534] [Q&A] [answered]
I have a nested matrix field and would like to get the owner of an entry a few levels deep. I understand I could do something like `{{ entry.owner.owner.title }}` etc, but is there a way to get the root entry directly in Twig?

**Comment by brandonkelly** [comment, marked as answer]
This will be possible in Craft 4.12 and 5.4, via `getRootOwner()`: (b1e4385de04b01d52ebdf28df58af1920296099a)

```twig
{{ entry.getRootOwner().title }}
```

**Reply by brandonkelly** [reply]
Craft 4.12.0 and 5.4.0 are out now with `Element::getRootOwner()`.

**Reply by shifuma** [reply]
This is great! Is there a way to use it with `relatedTo` or perhaps something like `rootOwner` on a query?

**Reply by brandonkelly** [reply]
Sure… what exactly are you trying to do?

**Reply by shifuma** [reply]
I have a 2 level deep nested Matrix:

- Entry
- - Matrix Entry
- - - Nested Matrix Entry

I would like to query for the 'Nested Matrix Entry' using the root entry ID. Something like this: `craft.entries().field('nestedMatrixField').rootOwner(entry.id).all()`

**Reply by brandonkelly** [reply]
Ah, sorry, no that’s not currently possible. You would need to fetch them in two steps – first grab the `Matrix Entry` entries, then their `Nested Matrix Entry` entries. You can use `eagerly()` on the nested query to reduce the overhead though:

```twig
{% set matrixEntries = entry.matrixEntries.all() %}
{% set nestedEntries = [] %}

{% for matrixEntry in matrixEntries %}
  {% set nestedEntries = nestedEntries|merge(matrixEntry.nestedMatrixEntries.eagerly().all()) %}
{% endfor %}
```

**Reply by shifuma** [reply]
Sorry to bring this up again, but I wonder if you'd consider adding the ability to query by `rootOwner`. You solution could probably work, but I think it my case it would be a bit excessive.

Some more information on my setup:

- Events have various relation and other fields (category, date, day of the week etc)
- Companies - a Matrix field on Events
- Presentations - a Matrix field on Companies

On a company's page, I am listing all of their presentations. However, I want to then filter those based on fields on Events. For example, from Company A's page, I want to see all of their presentations made at Events on a Saturday. That's where I thought this would be perfect.

```
-- `companies/_entry.twg` --
{% set events = craft.entries().section('events').dayOfTheWeek('Saturday').ids()
{% set presentations = craft.entries().field('presentations').rootOwner(events).all() %}
```

So whilst your solution could probably work, I think it the drawbacks are probably bigger than the benefits -- especially when I have 5+ filters I want to filter events by.

**Reply by brandonkelly** [reply]
@shifuma Sorry, that’s not possible, as we don’t currently store elements’ root owner IDs in the database. Feel free to post that as a new discussion though, and if the idea gets traction we could consider making it possible down the road.
